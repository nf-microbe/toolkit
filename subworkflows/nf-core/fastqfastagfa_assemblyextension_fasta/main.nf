// Import subworkflows
include { FASTQFASTA_COBRA_FASTA        } from '../fastqfasta_cobra_fasta'
include { FASTQGFA_PHABLES_FASTA        } from '../fastqgfa_phables_fasta'
include { getWorkDirs; rmEmptyFastAs    } from '../utils_nfmicrobe_functions/main'

workflow FASTQFASTAGFA_ASSEMBLYEXTENSION_FASTA {

    take:
    fastq_gz            // channel: [ [ meta.id, meta.single_end ], [ path(fastq_1), path(fastq_2) ] ]
    run_cobra           // boolean: false
    fasta_gz            // channel: [ [ meta.id, meta.single_end, meta.assembler ], path(fasta)]
    assembly_logs       // channel: [ [ meta.id, meta.single_end, meta.assembler ], path(logs) ]
    query_contigs_tsv   // channel: [ [ meta.id, meta.single_end, meta.assembler ], path(queries) ]
    run_phables         // boolean: false
    gfa_gz              // channel: [ [ meta.id ], gfa.gz ]
    phables_config      // channel: [ phables.config ]
    phables_db          // channel: [ phables_db ]

    main:

    ch_versions                     = Channel.empty()
    ch_extended_prefilt_fasta_gz    = Channel.empty()

    if (run_cobra) {
        //
        // SUBWORKFLOW: Extend assemblies with cobra
        //
        FASTQFASTA_COBRA_FASTA(
            fastq_gz,
            fasta_gz,
            query_contigs_tsv,
            assembly_logs
        )
        ch_extended_prefilt_fasta_gz    = ch_extended_prefilt_fasta_gz.mix(FASTQFASTA_COBRA_FASTA.out.fasta_gz)
        ch_versions                     = ch_versions.mix(FASTQFASTA_COBRA_FASTA.out.versions.first())
    }

    if (run_phables) {
        //
        // SUBWORKFLOW: Extend assemblies with phables
        //
        FASTQGFA_PHABLES_FASTA(
            fastq_gz,
            gfa_gz,
            phables_config,
            phables_db
        )
        ch_extended_prefilt_fasta_gz    = ch_extended_prefilt_fasta_gz.mix(FASTQGFA_PHABLES_FASTA.out.fasta_gz)
        ch_versions                     = ch_versions.mix(FASTQGFA_PHABLES_FASTA.out.versions.first())
    }

    // REMOVE EMPTY FASTA FILES FROM CHANNEL
    ch_extended_fasta_gz = rmEmptyFastAs(ch_extended_prefilt_fasta_gz, false)

    emit:
    fasta_gz    = ch_extended_fasta_gz  // channel: [ val(meta), [ fasta.gz ] ]
    versions    = ch_versions           // channel: [ versions.yml ]
}

