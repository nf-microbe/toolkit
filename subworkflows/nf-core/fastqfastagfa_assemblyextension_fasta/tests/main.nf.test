nextflow_workflow {

    name "Test Subworkflow FASTQFASTAGFA_ASSEMBLYEXTENSION_FASTA"
    script "../main.nf"
    workflow "FASTQFASTAGFA_ASSEMBLYEXTENSION_FASTA"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastqfastagfa_assemblyextension_fasta"
    tag "subworkflows/fastqfasta_cobra_fasta"
    tag "subworkflows/fastqgfa_phables_fasta"
    tag "subworkflows/utils_nfmicrobe_functions"

    test("lambda - fastq.gz - fasta.gz - gfa.gz") {
        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_2.fastq.gz", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_copy_1.fastq.gz", checkIfExists: true),
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_copy_2.fastq.gz", checkIfExists: true),
                        ]
                    ]
                )
                input[1] = true
                input[2] = Channel.of(
                    [
                        [ id: 'test1', assembler: 'megahit' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fasta/SRR24085752.fa.gz", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2', assembler: 'megahit' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fasta/SRR24085752_copy.fa.gz", checkIfExists: true),
                        ]
                    ]
                )
                input[3] = Channel.of(
                    [
                        [ id: 'test1' ],
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/log/SRR24085752.megahit.log", checkIfExists: true)
                    ],
                    [
                        [ id: 'test2' ],
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/log/SRR24085752_copy.megahit.log", checkIfExists: true)
                    ]
                )
                input[4] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/text/SRR24085752_queries.tsv", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/text/SRR24085752_copy_queries.tsv", checkIfExists: true),
                        ]
                    ]
                )
                input[5] = true
                input[6] = Channel.of(
                    [
                        [ id: 'test1' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/gfa/SRR24085752.gfa.gz", checkIfExists: true),
                        ]
                    ],
                    [
                        [ id: 'test2' ],
                        [
                            file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/gfa/SRR24085752_copy.gfa.gz", checkIfExists: true),
                        ]
                    ]
                )
                input[7] = file("${projectDir}/assets/configs/phables/phables_config.yml", checkIfExists: true)
                input[8] = null
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
