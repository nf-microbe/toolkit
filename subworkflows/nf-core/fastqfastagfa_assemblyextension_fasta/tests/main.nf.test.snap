{
    "lambda - fastq.gz - fasta.gz - gfa.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "extension": "phables"
                        },
                        "test1.phables.fasta.gz:md5,f8002e736576e646d5204488a7b32ec4"
                    ],
                    [
                        {
                            "id": "test2",
                            "extension": "phables"
                        },
                        "test2.phables.fasta.gz:md5,9c1b16b637bb6944107d382401801a4a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,30e3cdccc896087caf1f8d734d03a63a",
                    "versions.yml:md5,b56e74776e8ee3e6c3d06622e91f28b8"
                ],
                "fasta_gz": [
                    [
                        {
                            "id": "test1",
                            "extension": "phables"
                        },
                        "test1.phables.fasta.gz:md5,f8002e736576e646d5204488a7b32ec4"
                    ],
                    [
                        {
                            "id": "test2",
                            "extension": "phables"
                        },
                        "test2.phables.fasta.gz:md5,9c1b16b637bb6944107d382401801a4a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,30e3cdccc896087caf1f8d734d03a63a",
                    "versions.yml:md5,b56e74776e8ee3e6c3d06622e91f28b8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-13T13:55:26.602849168"
    }
}